### PR TITLE
feat(lora): HuggingFace model support (--lora-rank was silently ignored for HF)

### PR DESCRIPTION
**Pull request by saforem2**
## The bug

`--lora-rank 16 --model meta-llama/Llama-3.2-1B` **silently did nothing.**

Not an error — the flag was accepted, the run logged normally, and you got a full fine-tune wearing a LoRA run's clothes: plausible loss curve, and a checkpoint ~100× larger than intended.

Two independent causes:

1. **The LoRA block was unreachable for HF models.** It sat inside the native model's `else:` branch in `fsdp_tp.train`, so the HF path never called `apply_lora` at all.
2. **The names wouldn't have matched anyway.** `lora.py` hardcoded `wq/wk/wv/wo` + `w1/w2/w3`; HF Llama spells the same roles `q_proj/k_proj/v_proj/o_proj` + `gate_proj/up_proj/down_proj` + `lm_head`. Zero overlap.

Same silent-failure class as the rendezvous leak fixed in #218 — everything *looks* fine, so nothing prompts you to check.

## The fix

`LoraConfig` already gates by **role** (`train_attn`/`train_mlp`/`train_unembed`), and the module docstring already argued why: *"the roles are stable across model families even when the attribute spellings are not."* The abstraction just stopped at its own front door. This carries it through.

| role | native | HuggingFace |
|---|---|---|
| `attn` | `wq` `wk` `wv` `wo` | `q_proj` `k_proj` `v_proj` `o_proj` |
| `mlp` | `w1` `w2` `w3` | `gate_proj` `up_proj` `down_proj` |
| `unembed` | `output` | `lm_head` |

`target_names()` emits **both** vocabularies rather than sniffing the architecture. That's safe rather than sloppy: `apply_lora` wraps a child only when its attribute name is in the target set **and** it is an `nn.Linear`, so a spelling the model doesn't use never fires. No detection heuristic, nothing to mis-detect.

Three things found by experiment, each of which would otherwise have shipped as a bug:

- **The unembed lookup must write back to the attribute it matched.** The old code did `model.output = LoRALinear(...)`; on an HF model that creates a *new* attribute and leaves `lm_head` a plain `Linear`, so the forward path never sees the adapter — it reports success while training nothing.
- **Tied embeddings are fine, and I'd initially assumed otherwise.** With `tie_word_embeddings=True`, `lm_head.weight` *is* `embed_tokens.weight`; the adapter is additive and the shared tensor stays frozen. A test pins that wrapping doesn't thaw the input embedding.
- **Fused-QKV families genuinely can't work this way.** `transformers.pytorch_utils.Conv1D` is **not** an `nn.Linear` subclass (verified), so GPT-2 is unwrappable whatever it's called; Falcon/GPT-NeoX pack q/k/v into one tensor. The error now names that limitation instead of asking whether the model is an ezpz `Transformer`.

## The guard

`apply_lora` already raised on a *total* miss. The dangerous case is a **partial** match — it returns happily having adapted less than asked. So after application, the finished module tree is inspected and a requested role that produced no adapters is refused at setup. The run also logs adapter count and trainable/total params, so a run that isn't actually low-rank is visible in the log rather than only in the checkpoint size.

## Tests

`tests/test_tinker_lora_hf.py` builds `LlamaForCausalLM` from an in-memory `LlamaConfig` — deliberately **not** from the `tiny-random-llama-2/` directory in the working tree, which is untracked and absent on a clean clone.

The gate throughout is that **every trainable parameter receives a gradient** after a real forward/backward. "It ran without raising" is not the bar: a LoRA branch detached from the graph runs perfectly well and trains nothing.

One test asserts the LoRA block's **indentation is 4 spaces**, which reads odd until you realize that's precisely the bug — one level deeper and it's back inside the native branch. Verified by re-nesting the block and confirming the test fails, then restoring and confirming it passes. A guard that can't fail isn't a guard.

## Verification

Local: **1581 passed, 11 skipped in 77s**, `ruff` clean.

On an HF Llama: 15 adapters including `lm_head`, **30/30** trainable params receive gradients, 61,760 of 229,520 params trainable.

ALCF hardware validation on Sunspot, Aurora and Polaris is running now; results will be posted to this PR before merge.

## Summary by Sourcery

Enable reliable LoRA fine-tuning for HuggingFace Llama-family models while failing clearly when requested adapters cannot be applied.

New Features:
- Support LoRA adapters for HuggingFace Llama-family causal language models using role-based targets across native and HF projection names.

Bug Fixes:
- Ensure --lora-rank is applied on the HuggingFace training path instead of being silently ignored.
- Wrap the matched HuggingFace lm_head correctly when adapting the unembedding layer.
- Reject partial LoRA applications and unsupported fused-QKV architectures instead of allowing silent no-ops.

Enhancements:
- Log the number of adapters and trainable versus total parameters for LoRA runs.
- Preserve correct behavior for tied embeddings while keeping shared base weights frozen.

Documentation:
- Document HuggingFace LoRA usage, role-to-module mappings, parameter reporting, supported model families, and architecture limitations.

Tests:
- Add coverage for HuggingFace projection wrapping, gradient flow, freezing, target selection, unembedding, tied embeddings, unsupported architectures, and validation of partial applications.

### CODE REVIEW

**Comment by saforem2** [comment]
## ALCF hardware validation — Sunspot ✅ Aurora ✅

Both Intel XPU systems pass on real hardware, running commit `1c03261`.

| | Sunspot | Aurora |
|---|---|---|
| job | `12473634` | `8774780` |
| torch | `2.13.0a0+gitcf30153` | `2.10.0a0+git449b176` |
| adapters | **196** | **196** |
| trainable | 10,092,544 / 606,142,464 (**1.665%**) | same |
| distributed run | 12 ranks, 5 iters | 12 ranks, 5 iters |
| loss | 12.69 → 11.63 ↓ | 12.69 → 11.44 ↓ |
| exit | 0 | 0 |

Model: `Qwen/Qwen3-0.6B` (fully offline — `HF_HUB_OFFLINE=1`), `--lora-rank 16 --tp 1 --dataset random`.

Two gates, because "the job exited 0" is not evidence:

1. **Adapters actually applied** — 196 modules wrapped and only 1.665% of parameters trainable. On `main` this same command wraps **zero** and trains all 606M.
2. **Real distributed training** — 12 ranks via `ezpz.launch`, loss decreasing across iterations. Confirmed from the run's own JSONL rather than the job log, since I'd piped the log through `tail` and clipped the LoRA line:

```
LoRA: 196 adapters (rank=16, targets=attn,mlp); 10,092,544 trainable of 606,142,464
```

### Polaris

Running (`7551793`). It needed two environment workarounds that are **not** related to this change, both consequences of the recent PE upgrade:

- `module load conda` is broken — the conda modules pin `gcc-native/14.2` and `cray-hdf5-parallel/1.14.3.5`, neither of which still exists. `ezpz_setup_conda_polaris` hard-pins `conda/2025-09-25`, so **`ezpz_setup_env` is affected on Polaris generally**, not just here. Filing separately.
- `cray-mpich` went 8.x → 9.1.0 and renamed the library. The venv's `mpi4py` dlopens `libmpi.so{,.12,.40}`; only `libmpi_gnu.so.12` exists now, so it failed at import. Worked around with a symlink shim.

### A note on the harness

My first three job submissions failed for a reason worth recording: PBS runs job scripts under a **non-login** shell where `module` is undefined and `MODULEPATH` is unset, so `module load frameworks` was a silent no-op and torch fell back to a build with no MKL. That is the same failure class as #216 — and sourcing lmod's init is *not* sufficient, because it supplies the `module` function without the site `MODULEPATH`. The scripts now re-exec once under `bash -l` and assert `MODULEPATH` is set before continuing.

`utils.sh` never checks `MODULEPATH`, which is a real gap in the #216 family. Also filing separately rather than folding it in here.

File: src/ezpz/examples/fsdp_tp.py [hunk 1]
```diff
@@ -2782,32 +2786,65 @@ def train(
...
+        # requested role cannot quietly become a no-op.
+        #
+        # (Tied embeddings are fine and deliberately not special-cased:
+        # `lm_head` is still an nn.Linear, the adapter is additive, and
+        # the shared weight stays frozen. Verified in
+        # tests/test_tinker_lora_hf.py.)
+        _applied = [n for n, _ in _lora.iter_lora_modules(model)]
+        if "unembed" in _targets and not any(
+            n.rsplit(".", 1)[-1] in _lora.UNEMBED_TARGETS for n in _applied
+        ):
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Confirmed and fixed in `6536cd4` — this was a real gap in the guard I added, and three reviewers flagged it independently.

I reproduced it before fixing: a module tree with `q_proj`/`k_proj` and no MLP, asked for `attn,mlp`, adapts 2 modules. `apply_lora` does not raise (attn matched), and the old guard only looked at `unembed`, so it passed — exactly the silent partial application the guard exists to prevent.

Now every requested role is validated against the adapted leaf names, in both native and HF spellings, and the error names which roles came up empty, what *was* adapted, and the attribute names each missing role expects.

Two tests pin it: one asserts all three role tables appear in the guard, and one is behavioural — builds the attn-only model above and confirms `mlp` is identified as unmatched.

File: src/ezpz/tinker/lora.py [hunk 2]
```diff
@@ -56,6 +56,27 @@
...
+# no detection heuristic and cannot mis-detect.
+#
+# Fused-QKV families are deliberately absent. GPT-2 projections are
+# `transformers.pytorch_utils.Conv1D`, which is NOT an `nn.Linear`
+# subclass, so `apply_lora` skips them whatever they are called; Falcon
+# and GPT-NeoX pack qkv into one `query_key_value` that would have to be
+# split before an adapter could target a single projection. Naming them
+# here would produce silent no-ops, so the guard in `fsdp_tp` reports them
+# as unsupported instead.
+HF_ATTN_TARGETS: tuple[str, ...] = ("q_proj", "k_proj", "v_proj", "o_proj")
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/tinker/lora.py:74]: Fixed in `6536cd4`. The module docstring now covers both `ezpz.models.llama` and HF Llama-family models.

File: src/ezpz/examples/fsdp_tp.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Confirmed and fixed in `6536cd4` — this was a real gap in the guard I added, and three reviewers flagged it independently.

I reproduced it before fixing: a module tree with `q_proj`/`k_proj` and no MLP, asked for `attn,mlp`, adapts 2 modules. `apply_lora` does not raise (attn matched), and the old guard only looked at `unembed`, so it passed — exactly the silent partial application the guard exists to prevent.

Now every requested role is validated against the adapted leaf names, in both native and HF spellings, and the error names which roles came up empty, what *was* adapted, and the attribute names each missing role expects.

Two tests pin it: one asserts all three role tables appear in the guard, and one is behavioural — builds the attn-only model above and confirms `mlp` is identified as unmatched.

File: tests/test_tinker_lora_hf.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+well and trains nothing, which is the exact failure being fixed.
+"""
+
+from __future__ import annotations
+
+import pytest
+
+torch = pytest.importorskip("torch")
+transformers = pytest.importorskip("transformers")
+
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_tinker_lora_hf.py:32]: Fixed in `6536cd4` — dropped the unused binding, now a bare `pytest.importorskip("transformers")`.

File: src/ezpz/examples/fsdp_tp.py [hunk 4]
```diff
@@ -2782,32 +2786,65 @@ def train(
...
+            raise SystemExit(
+                "--lora-target includes 'unembed' but no unembedding "
+                f"projection was adapted on {type(model).__name__}: "
+                f"expected one of {list(_lora.UNEMBED_TARGETS)} as a "
+                "top-level nn.Linear. Drop 'unembed' from --lora-target, "
+                "or use a model that exposes a separate output head."
             )
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:2854]: Confirmed and fixed in `6536cd4` — this was a real gap in the guard I added, and three reviewers flagged it independently.

I reproduced it before fixing: a module tree with `q_proj`/`k_proj` and no MLP, asked for `attn,mlp`, adapts 2 modules. `apply_lora` does not raise (attn matched), and the old guard only looked at `unembed`, so it passed — exactly the silent partial application the guard exists to prevent.

Now every requested role is validated against the adapted leaf names, in both native and HF spellings, and the error names which roles came up empty, what *was* adapted, and the attribute names each missing role expects.

Two tests pin it: one asserts all three role tables appear in the guard, and one is behavioural — builds the attn-only model above and confirms `mlp` is identified as unmatched.

File: src/ezpz/tinker/lora.py [hunk 5]
```diff
@@ -97,11 +118,19 @@ def __post_init__(self) -> None:
             )
 
     def target_names(self) -> tuple[str, ...]:
+        """Attribute names to adapt, in both native and HF spellings.
+
+        Emitting both is safe because `apply_lora` matches on the
+        attribute name AND requires an `nn.Linear`, so the spellings a
+        given model does not use never fire.
+        """
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/tinker/lora.py]: Fixed in `6536cd4`. `target_names()` now says it covers `train_attn`/`train_mlp` only, and explains why `train_unembed` is absent: `apply_lora` resolves it against the ROOT module via `UNEMBED_TARGETS`, so listing `output`/`lm_head` in the generic set would wrap any nested child that happened to share the name.

**Comment by saforem2** [comment]
## Review round addressed (`6536cd4`)

Three reviewers independently found the same real gap in the guard I'd added, and they were right.

**The guard only validated `unembed`.** A model matching `attn` but with no recognizable MLP passed `--lora-target attn,mlp` having adapted less than asked — the exact silent partial application the guard exists to prevent. I reproduced it before fixing:

```
requested attn+mlp, adapted: {'k_proj', 'q_proj'}
mlp got ZERO adapters -> True
...and apply_lora did NOT raise (attn matched)
```

Now every requested role is checked against the adapted leaf names in both spellings, and the error names which roles came up empty, what *was* adapted, and the attribute names each missing role expects. Two tests pin it: a structural one covering all three role tables, and a behavioural one that builds the attn-only model above and confirms `mlp` is flagged.

Also from the same round: the module docstring still said `ezpz.models.llama` only; `target_names()` didn't explain why `train_unembed` is excluded (it resolves against the *root* module, so listing `output`/`lm_head` in the generic set would wrap any nested child sharing the name); and one unused `importorskip` binding.

All six threads replied to and resolved. Suite: **1583 passed, 11 skipped**, `ruff` clean.

## Polaris

Still working on it, and it is **entirely environmental** — no code issue.

The eagle venv's `mpi4py` dlopens `libmpi.so{,.12,.40}`, none of which survived the cray-mpich 8.x → 9.1.0 rename. A symlink shim to `libmpi_gnu.so.12` fixed it **on the login node**, but the compute node resolved `MPICH_LIB=` empty — `/opt/cray/pe/mpich/9.1.0/ofi/gnu/*/lib` does not exist there, so the symlinks pointed at nothing. A probe job is out to find the real compute-node path.

Sunspot and Aurora are already green (196 adapters, 1.665% trainable, 12 ranks, loss decreasing), so the feature itself is validated on Intel XPU; Polaris adds the CUDA/nccl path.
